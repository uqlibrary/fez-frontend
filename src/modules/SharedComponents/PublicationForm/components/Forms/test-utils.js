import { useForm } from 'react-hook-form';
import { render, WithReduxStore } from 'test-utils';
import React from 'react';
import { formValues } from '../PublicationForm';

export const setupWrapper = (Form, props) => {
    const Wrapper = () => {
        const {
            control,
            formState: { isSubmitting },
        } = useForm();

        // eslint-disable-next-line react/prop-types
        return <Form {...{ control, isSubmitting, ...props, formValues: formValues(props?.formValues || {}) }} />;
    };

    return render(
        <WithReduxStore>
            <Wrapper />
        </WithReduxStore>,
    );
};
