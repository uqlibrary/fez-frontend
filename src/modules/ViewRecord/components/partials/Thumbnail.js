import React from 'react';
import PropTypes from 'prop-types';
import { styled } from '@mui/material/styles';

import ExternalLink from 'modules/SharedComponents/ExternalLink/components/ExternalLink';
import BrokenImage from '@mui/icons-material/BrokenImage';
import Lock from '@mui/icons-material/Lock';
import locale from 'locale/pages';
import { Img } from 'react-image';
import CircularProgress from '@mui/material/CircularProgress';

export const getTestId = filename => `preview-link-${filename}`;

const StyledThumbnailImage = styled(Img)(() => ({
    width: '100%',
    '&:hover': {
        cursor: 'pointer',
    },
}));

const Thumbnail = ({
    mediaUrl,
    webMediaUrl,
    previewMediaUrl,
    thumbnailMediaUrl,
    thumbnailFileName,
    securityStatus,
    fileName,
    mimeType,
    onClick,
    checksums,
}) => {
    const txt = locale.pages.viewRecord;

    if (
        (fileName && mimeType.indexOf('pdf') >= 0) ||
        (mimeType.indexOf('octet-stream') >= 0 && mediaUrl.indexOf('flv') >= 0)
    ) {
        return (
            <ExternalLink href={mediaUrl} title={fileName} openInNewIcon={false} id="thumbnail">
                <StyledThumbnailImage
                    crossorigin="anonymous"
                    src={thumbnailMediaUrl}
                    alt={thumbnailFileName}
                    loader={<CircularProgress size={15} thickness={1} />}
                    unloader={<BrokenImage color={'secondary'} data-testid={`${fileName}-error`} />}
                />
            </ExternalLink>
        );
    }
    return (
        <a
            onClick={e => {
                e.preventDefault?.();
                onClick({ fileName, mediaUrl, previewMediaUrl, mimeType, webMediaUrl, securityStatus, checksums });
            }}
            onKeyPress={e => {
                e.preventDefault?.();
                onClick({ fileName, mediaUrl, previewMediaUrl, mimeType, webMediaUrl, securityStatus, checksums });
            }}
            title={mediaUrl && txt.thumbnailTitle.replace('[image]', mediaUrl)}
            id={getTestId(fileName)}
            data-testid={getTestId(fileName)}
        >
            {securityStatus ? (
                <StyledThumbnailImage
                    src={thumbnailMediaUrl}
                    alt={thumbnailFileName}
                    loader={<CircularProgress size={15} thickness={1} />}
                    unloader={<BrokenImage color={'secondary'} />}
                />
            ) : (
                <Lock color={'secondary'} sx={{ opacity: 0.5 }} />
            )}
        </a>
    );
};
Thumbnail.propTypes = {
    mediaUrl: PropTypes.string,
    webMediaUrl: PropTypes.string,
    previewMediaUrl: PropTypes.string,
    thumbnailMediaUrl: PropTypes.string,
    thumbnailFileName: PropTypes.string,
    securityStatus: PropTypes.bool,
    fileName: PropTypes.string,
    mimeType: PropTypes.string.isRequired,
    onClick: PropTypes.func,
    checksums: PropTypes.object,
};

export default Thumbnail;
