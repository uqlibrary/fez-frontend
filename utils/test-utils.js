/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';

import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';

import mediaQuery from 'css-mediaquery';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');
const mime = require('mime-types');

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@testing-library/dom';
import preview from 'jest-preview';

export const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <LocalizationProvider dateAdapter={AdapterMoment}>{props.children}</LocalizationProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => {
    return {
        user: userEvent.setup(),
        ...render(ui, { wrapper: AllTheProviders, ...options }),
    };
};

export const WithRouter = ({ children, route = '/', initialEntries = [route] }) => {
    const routes = [{ path: route, element: children }];
    const router = createMemoryRouter(routes, { initialEntries: initialEntries });
    return <RouterProvider router={router} />;
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

export const assertTooltipText = async (trigger, tooltipText) => {
    expect(trigger).toBeInTheDocument();
    await userEvent.hover(trigger);
    await waitFor(() => {
        expect(screen.getByRole('tooltip', { name: tooltipText, hidden: true })).toBeVisible();
    });
};

const extensionToMimeMap = {
    tiff: 'image/tiff',
    wav: 'audio/wave',
    zip: 'application/zip',
};
const mimeType = filename => extensionToMimeMap[filename.split('.').pop()];

export const defaultDatastream = {
    dsi_embargo_date: null,
    dsi_open_access: 1,
    dsi_label: '',
    dsi_copyright: null,
    dsi_state: 'A',
    dsi_size: 27932352,
};

export const createFezDatastreamInfoArray = (datastreams, pid = null, withPreview = true) => {
    let processed = datastreams.filter(datastream => typeof datastream === 'object' || typeof datastream === 'string');
    if (withPreview) {
        processed = [];
        datastreams.forEach(datastream => {
            processed.push(datastream);
            const filename = typeof datastream === 'object' ? datastream.dsi_dsid : datastream;
            const mimetype = mimeType(filename);
            // bail in case it's not derivable
            if (
                !filename.includes('.pdf') &&
                !mimetype.includes('image') &&
                !mimetype.includes('audio') &&
                !mimetype.includes('video')
            ) {
                return;
            }
            // eslint-disable-next-line no-nested-ternary
            const derivativeMimetype = mimetype.includes('audio')
                ? 'audio/mp3'
                : mimetype.includes('video')
                ? 'video/mp4'
                : 'image/jpg';

            (derivativeMimetype.includes('image') ? ['preview_', 'thumbnail_', 'web_'] : ['']).forEach(prefix => {
                processed.push({
                    dsi_dsid: `${prefix}${filename.split('.')[0]}_t.${derivativeMimetype.split('/').pop()}`,
                    dsi_mimetype: derivativeMimetype,
                });
            });
        });
    }

    return processed.map((datastream, index) => ({
        ...defaultDatastream,
        ...{
            dsi_id: index + 1,
            dsi_pid: pid,
            dsi_order: index + 1,
            dsi_mimetype: mime.lookup(typeof datastream === 'object' ? datastream.dsi_dsid : datastream),
        },
        ...(typeof datastream === 'object'
            ? datastream
            : {
                  dsi_dsid: datastream,
              }),
    }));
};

export const withDatastreams = (sources, datastreams, callback) =>
    datastreams.map(item => {
        const source = sources.find(datastream =>
            datastream.dsi_dsid.includes(
                item.dsi_dsid
                    .replace('_t.', '.')
                    .replace('preview_', '')
                    .replace('thumbnail_', '')
                    .replace('web_', '')
                    .split('.')[0],
            ),
        );
        const isDerivative = item.dsi_dsid !== source.dsi_dsid;
        callback(item, source, isDerivative);
    });

export const getDatastreamByFilename = (filename, datastreams) =>
    datastreams.find(datastream => datastream.dsi_dsid === filename);

export const createMatchMedia = width => {
    return query => ({
        matches: mediaQuery.match(query, { width }),
        /* istanbul ignore next */
        addListener: () => {},
        /* istanbul ignore next */
        removeListener: () => {},
    });
};

export const getFilenameExtension = filename => filename.split('.').pop();
export const getFilenameBasename = filename =>
    filename.replace(new RegExp(`/\.${getFilenameExtension(filename)}$/`), '');
export const addFilesToFileUploader = files => {
    const { screen, fireEvent } = reactTestingLib;
    // create a list of Files
    const fileList = files.map(file => {
        // if file it's a string, treat it as a filename
        if (typeof file === 'string') {
            return new File([getFilenameBasename(file)], file, { type: `image/${getFilenameExtension(file)}` });
        }
        // otherwise expect it to be a object with filename and mimeType keys
        return new File([getFilenameBasename(file.filename)], file.filename, { type: file.mimeType });
    });
    // drag and drop files
    fireEvent.drop(screen.getByTestId('fez-datastream-info-input'), {
        dataTransfer: {
            files: fileList,
            types: ['Files'],
        },
    });
};
export const setFileUploaderFilesToClosedAccess = async (files, timeout = 500) => {
    const { fireEvent } = reactTestingLib;
    // set all files to closed access
    for (const file of files) {
        const index = files.indexOf(file);
        await waitFor(() => screen.getByText(new RegExp(getFilenameBasename(file))), { timeout });
        fireEvent.mouseDown(screen.getByTestId(`dsi-open-access-${index}-select`));
        fireEvent.click(screen.getByRole('option', { name: 'Closed Access' }));
    }
};

export const assertEnabled = (resolver, ...items) => items.forEach(item => expect(resolver(item)).toBeEnabled());
export const assertDisabled = (resolver, ...items) => items.forEach(item => expect(resolver(item)).toBeDisabled());

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    withRedux,
    AllTheProviders,
    WithReduxStore,
    assertTooltipText,
    WithRouter,
    createFezDatastreamInfoArray,
    getDatastreamByFilename,
    withDatastreams,
    createMatchMedia,
    preview,
    userEvent,
    assertEnabled,
    assertDisabled,
    getFilenameExtension,
    getFilenameBasename,
    addFilesToFileUploader,
    setFileUploaderFilesToClosedAccess,
};
