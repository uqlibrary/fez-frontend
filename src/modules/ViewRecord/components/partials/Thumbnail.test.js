import React from 'react';
import Thumbnail from './Thumbnail';
import { rtlRender, fireEvent, within } from 'test-utils';

function setup(testProps = {}) {
    const props = {
        mediaUrl: '/test/mediaUrl',
        previewMediaUrl: '/test/preview/mediaUrl',
        thumbnailMediaUrl: '/test/thumbnail/mediaUrl',
        thumbnailFileName: 'thumbnail.jpg',
        mimeType: 'image/jpeg',
        onClick: jest.fn(),
        classes: {
            image: '',
        },
        ...testProps,
    };
    return rtlRender(<Thumbnail {...props} />);
}

describe('Thumbnail component', () => {
    it('should render component', () => {
        const { container } = setup();
        expect(container).toMatchSnapshot();
    });

    it('should run onClick function on click', () => {
        const onClick = jest.fn();
        const { getByTitle } = setup({ onClick: onClick });
        fireEvent.click(getByTitle('Click to open a preview of /test/mediaUrl'));
        expect(onClick).toHaveBeenCalledTimes(1);
        fireEvent.keyPress(getByTitle('Click to open a preview of /test/mediaUrl'), { key: 'Enter', keyCode: 13 });
        expect(onClick).toHaveBeenCalledTimes(2);
    });

    it('should render a lock icon with empty security status', () => {
        const { getByTestId } = setup();
        expect(getByTestId('LockIcon')).toBeInTheDocument();
    });

    it('should render image with security status', () => {
        const { queryByTestId, getByRole } = setup({ securityStatus: true });
        expect(getByRole('progressbar')).toBeInTheDocument();
        expect(queryByTestId('LockIcon')).not.toBeInTheDocument();
    });
    it('should render external link with a progress bar for FLV files', () => {
        const { getByRole } = setup({
            fileName: 'video.mov',
            mediaUrl: 'video.flv',
            mimeType: 'application/octet-stream',
        });

        const element = getByRole('link', { title: 'video.mov' });
        expect(element).toHaveAttribute('href', 'video.flv');
        expect(within(element).getByRole('progressbar')).toBeInTheDocument();
    });

    it('should render PDF files as a thumbnail only plus link to original', () => {
        const { getByRole } = setup({
            fileName: 'original.pdf',
            mediaUrl: 'pdf.jpg',
            thumbnailFileName: 'thumbnail.jpg',
            mimeType: 'application/pdf',
        });
        const element = getByRole('link', { title: 'original.pdf' });
        expect(element).toHaveAttribute('href', 'pdf.jpg');
        expect(within(element).getByRole('progressbar')).toBeInTheDocument();
    });
});
