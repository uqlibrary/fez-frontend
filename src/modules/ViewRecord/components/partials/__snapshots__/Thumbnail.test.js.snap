// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumbnail component should render component 1`] = `
<div>
  <a
    data-testid="preview-link-undefined"
    id="preview-link-undefined"
    title="Click to open a preview of /test/mediaUrl"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeMedium css-1l910iq-MuiSvgIcon-root"
      data-testid="LockIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2m-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1z"
      />
    </svg>
  </a>
</div>
`;
